Golem Rework

Jacquerel opened the pull request:
## About The Pull Request

Fixes #74375 *
Fixes #73235
Fixes #64420 *

*Technically, by making these things either intentional or nonexistent

This PR implements this design document: https://hackmd.io/@Y6uzGFDGSXKRaWDNicSiEg/BkRr176st
Put briefly, this will remove every existing golem subtype and consolidate golems into a single species with cool new sprites.
NOT implemented from that PR is the ability to eat Telecrystals, I couldn't come up with an appropriate visual that can stack with the existing ones, but that should be a reasonably trivial add for a future artist & developer.

![image](https://user-images.githubusercontent.com/7483112/235223788-cbfda4bc-1b4e-4613-b352-6179decb757b.png)
The diamond one renders you invisible so uh... hard to get a screenshot of the sprite.

New Golems have a food-based mechanic where their hunger decays pretty quickly and can only be replenished by eating minerals. They start moving slower as they get hungrier, until eventually they become completely immobilised and need to be rescued.
Eating different kinds of minerals will visually change your sprite and give you a special effect in a similar way to old golems, but temporary. While transformed, you can't eat any other kind of mineral which would transform you (but can still consume glass).
To see the full list of effects, look at the hackmd above.

![image](https://user-images.githubusercontent.com/7483112/235224479-86f57ca2-4363-42de-8532-2a1141df09c9.png)

In service of these sprites working I have refactored the `species/offset_features` feature by killing it and delegating that responsibility to limbs instead. Rather than applying an offset to items due to your species, it is due to your weird head or arms. This makes overall more sense to me, but it inflates the code changes in this PR somewhat.
It doesn't make a lot of sense to atomise unfortunately because that code also seemed to be entirely unused until I tried to use it in this PR, so you wouldn't be able to tell if my changes broke anything. I might make a downstream sad by doing this.

All of the actual numbers in this PR are made up and only loosely tested, it will need some testmerges to gather feedback about whether it sucks or not.

TODO:

- [x] Actually implement the rest of the transformation effects.
- [x] Add status alerts with tooltips to tell you what's happening to you.
- [x] Add examine tooltips to edible minerals if you have a golem brain.
- [x] Possibly add examine tooltips to the Golem to let other people know what their active effect is doing?
- [x] Check if a limb we have is actually a golem one before drawing rocks growing out of it.
- [x] Make Golems become immobilised if they are starving.
- [x] Test golems wearing clothes.
- [x] Test humans with golem limbs (Test conclusion: it's fucked, don't allow it)
- [x] Let golems smelt materials if they are stood in lava.
- [x] Figure out what to do with the couple of places which used to spawn specific golem subtypes.
- [x] Maybe let them mine by punching or tie that to one of the mineral buffs idk.
- [ ] Add excitement to any effects which don't seem useful or interesting in practice.
- [ ] Tune values so that you both have to actively care about having food on hand but also can meaningfully enjoy the game.

Other relevant changes:
I reworked how bioscrambling works based off bodypart bodytypes, to automatically exclude golem limbs in either direction. There's really no way to have those work on humans or vice versa. Organs still fly though.

## Why It's Good For The Game

Reduces the pollution of the codebase by tens of subtypes of one gimmick species.
Refocuses golems to be good at mining, and able to perform secondary roles based on what they are being supplied with to eat, instead of being a huge number of visually similar guys who mostly just hung around being humans anyway.
These sprites look really good (but I need to make sure they work well when put on other carbons, and with clothes).

## Changelog

:cl: Jacquerel with new icons mostly by INFRARED_BARON 
balance: Golems have been reworked into a hunger-based species who gain temporary transformations when eating different kinds of mineral.
del: The previous golem subtypes no longer exist.
/:cl:

Jacquerel committed:
Removes all the subtypes

Jacquerel committed:
Some leftovers
Don't worry comments, we did it

Jacquerel committed:
Adds sprites

Jacquerel committed:
Golems can eat any stack (for now) and not food
Golems get slow if they're hungry

Jacquerel committed:
Corrects stomach physiology
Golem darkvision

Jacquerel committed:
Golems can see in the dark & pulse ores
Eat one stack per bite

Jacquerel committed:
You can get buffs from certain minerals

Jacquerel committed:
Fixes plasteel path.

Jacquerel committed:
Fix missing args in extended procs
Remove unused variable

Jacquerel commented:
Of course please review whatever but specifically this part is the one I'm worried about not wanting to get too deep into before I have to refactor it to work completely differently:
https://github.com/tgstation/tgstation/pull/74197/files#diff-a647b4ae7435451897dfa6a05b9fc3ba0ac5bff2ff3c4af2cd17a5fe633e21c2

Jacquerel committed:
Probably don't need this

Sealed101 commented:
how would (or should) nu-golems interact with stabilized silvers (reduces hunger rate) or consuming crossbreeds' slime cookies (which aren't even a food and i think they add nutrition dierctly alongside a special effect)?

Jacquerel committed:
Were there really 26 of these fucking hell

private-tristan commented:
Will disconnected golems need to be saved? Or will their hunger stop after some time?

Jacquerel commented:
> Will disconnected golems need to be saved? Or will their hunger stop after some time?

Probably the former, I can add some afk protection if this turns out to be an issue.

ATH1909 commented:
I'm not entirely a fan of it leading to total immobilization, but I'm also not a fan of gbj that isn't directly caused by other players in general.

> While transformed, you can't eat any other kind of mineral which would transform you.

So you just have to... wait to starve if you want to change golem species?

Jacquerel commented:
> > While transformed, you can't eat any other kind of mineral which would transform you.
> 
> So you just have to... wait to starve if you want to change golem type?

There are various things which _don't_ provide buffs which you can eat with a different effect active.
Eating (for example) uranium while you already have the uranium effect active doesn't reset the effect duration, it will expire 5 minutes after you gained it.

Elli-Skala commented:
> until eventually they become completely immobilised and need to be rescued.

This sounds quite awful being stuck in a gay baby jail to be completely honest (say a radioless golem being stuck somewhere remote unable to even end it outside ghosting) 

The idea of a golem going motionless in a death state is abit interesting (say going into crit as a golem makes you pause in place and crit is your mind fading away) but being alive during it would be much akin to being stuck inside your body when you die

ATH1909 commented:
> > > While transformed, you can't eat any other kind of mineral which would transform you.
> > 
> > 
> > So you just have to... wait to starve if you want to change golem type?
> 
> There are various things which _don't_ provide buffs which you can eat with a different effect active. Eating (for example) uranium while you already have the uranium effect active doesn't reset the effect duration, it will expire 5 minutes after you gained it.

No, I mean, if you want to switch which golem effect you have active, you have to wait to starve enough to be able to eat a new kind of food? You can't end your existing effect early by eating new food?

Also, uh, being unable to renew a golem effect while it's active (leaving repeated gaps where you're just a normal dude, even if you have a massive stockpile of food on you) sounds like it'd cause awful gameplay feel. It also restricts golem effect design, because, say, immunity to space/fire/whatever needs to be on 24/7 without gaps or you're just kind of screwed.

Jacquerel committed:
Don't apply golem overlays to a human arm
Immobilise when starved
Actually allow people to feed golems

Jacquerel commented:
I don't know why you keep referencing "needing to starve yourself" because you don't, you just wait 5 minutes. 
I assume the set timer is so that you're committing to something for a while and to ensure that you don't commit to something without having to have prepared to deal with it for a while.
If you go into space and don't bring any plasteel with you for some reason then that's the equivalent of going into space with half an emergency oxygen tank. You'll survive two seconds of the buff falling off to eat another one, you're not "screwed" if it's not up 100% of the time actually.
(Also golems are just spaceproof at all times right now anyway but that'll probably change if we want it to be tied to Plasteel).

Additionally, these are the requirements of Oranges' design doc.
I don't say that to mean that they are written in stone but they _are_ unlikely to change before being playtested.
Maybe we'll end up making statue-ing last for a minute before generating some nutrition in your stomach or something, who knows.

ATH1909 commented:
> You'll survive two seconds of the buff falling off to eat another one, you're not "screwed" if it's not up 100% of the time actually.

I guess I just don't get why it can't be a refreshable buff instead of one that blocks you from eating that material again until it wears off. Like, it doesn't have to be a _stockpilable_ buff, just one that you can reset to its initial duration before it's run out.

Jacquerel committed:
Add 30 seconds per ore consumed up to 5 minutes

Jacquerel committed:
Status effect fluff
Golems can smelt ore using their appendix
Fixes merging inhand stacks with put_in_hands

Jacquerel committed:
Correct argument
Maybe satisfies unit test

Jacquerel committed:
Merge branch 'master' into subtypes-begone

Jacquerel committed:
i summon the red x

Jacquerel committed:
Merge branch 'master' into subtypes-begone

Jacquerel committed:
Temperature defines

Jacquerel committed:
Merge branch 'master' into subtypes-begone

Jacquerel committed:
Delete this

Jacquerel committed:
Plasteel, Gold, Plasma

Jacquerel committed:
Diamond

Jacquerel committed:
Titanium, Bananium

Jacquerel committed:
yeah I eat lightbulbs, what of it?

Jacquerel committed:
Gibtonite, i stepped away from this one and came back hours later and now i feel like i forgot something
well its probably not important

Jacquerel committed:
Fucking around with numbers

Jacquerel committed:
Bluespace part I

Jacquerel committed:
Shout out in the comments anyone who is reading my commit messages
Bluespace part II

optimumtact commented:
>I guess I just don't get why it can't be a refreshable buff 

because that's not how I designed it

Jacquerel committed:
Merge branch 'master' into subtypes-begone

Jacquerel committed:
Merge branch 'master' into subtypes-begone

Jacquerel committed:
Singleton food datums
Sound for gibtonite hand

Jacquerel committed:
Status alert for being statued

Jacquerel commented:
I'm moving house but I WILL get back to this after that (probably in time for the stale bot not to close it at least)

ATH1909 commented:
> > I guess I just don't get why it can't be a refreshable buff
> 
> because that's not how I designed it, it's forcing you into a choice you make, so that you can suffer whatever the consequences of that is, without just backing out of your choice with no repercussions.

to be clear, what I'm suggesting is making it so that if you're, say, a gold golem with 2 minutes remaining on your gold buff, you can chomp another gold bar to get brought back up to your gold buff's initial duration. I'm not arguing for making eating a diamond while you're a gold golem immediately switch you to the diamond buff or something like that.

optimumtact commented:
no

Jacquerel committed:
Merge branch 'master' into subtypes-begone

Jacquerel committed:
Add alerts with tooltips, which indicate when your food buff will soon expire

Jacquerel committed:
Extract golem edibility into a component
People with golem brains can now see if things are edible by examining them
You can see what a golem has eaten by looking at them

Jacquerel committed:
God damn I didn't realise you were going to post to complain about it every single push christ

Jacquerel committed:
Block golem limbs from being bioscrambled

Jacquerel committed:
There's too many assumptions attached to "not being organic" right now
make sure you can glue golem arms back onto golems

Jacquerel committed:
Apply bodypart traits when applying limb owner so it works properly for spawning with them rather than just gaining them (not many bodyparts have them i guess)
Start trying to make clothing offsets (I can already tell this feature hasnt been used much because it doesnt work)

Jacquerel committed:
Make golems render eyes if they gain them, but golem eyes are internal
Set golem species offsets
horrible evil code to mirror x offset by dir i hope someone has a better idea

Jacquerel commented:
why the hell does that screenshot have eyes

Jacquerel committed:
Remove eyehole sprites from golems

Jacquerel committed:
Quit your whining icebox
Updated screenshot

Jacquerel committed:
Finish up todos
Make golem shells & spawners work

Jacquerel committed:
Golems can mine with their bare hands

Jacquerel committed:
Merge branch 'master' into subtypes-begone

Jacquerel commented:
This is so close to done but `offset_features` is so worthless that i just need to rewrite it 😭

Jacquerel committed:
Kill species feature offset assoc list, delegate it to a datum
The amount of proc calls seems kind of cursed but it works now and didnt before

Jacquerel committed:
Merge remote-tracking branch 'tgstation/master' into subtypes-begone

Jacquerel committed:
No cheese

Jacquerel commented:
I think this actually might be ready for reviews now, although probably we need some playtesting in terms of "is the value of food vs hunger set correctly and are those mechanics sustainable".

san7890 commented on code/datums/components/food/golem_food.dm:
```diff
@@ -0,0 +1,112 @@
+		return COMPONENT_CANCEL_ATTACK_CHAIN
+	if (!golem_snack)
+		golem_snack = new(
+			/* loc = */ null,
+			/* name = */ source.name,
+			/* consume_food = */ consume_on_eat,
+			/* food_buff = */ snack_type,
+			/* owner = */ parent,
```
why have the args formatted like this, do named args not work on new()? if they don't this is fine but i've never seen this patterh before lol

Jacquerel replied:
Afraid so.

JohnFulpWillard replied:
I think this is mentioned somewhere in our standards as an alternative for this case, or at least I remember it being written somewhere...

san7890 commented on code/modules/mob/living/carbon/human/human_update_icons.dm:
```diff
@@ -189,17 +188,23 @@ There are several things that need to be remembered:
 
-	if(!get_bodypart(BODY_ZONE_HEAD)) //decapitated
+	var/obj/item/bodypart/head/my_head = get_bodypart(BODY_ZONE_HEAD)
+	if(!my_head) //decapitated
```
`isnull()` here?

san7890 commented on code/modules/mob/living/carbon/human/human_update_icons.dm:
```diff
@@ -216,17 +221,16 @@ There are several things that need to be remembered:
 
-	if(!get_bodypart(BODY_ZONE_HEAD)) //decapitated
+	var/obj/item/bodypart/head/my_head = get_bodypart(BODY_ZONE_HEAD)
+	if(!my_head) //decapitated
```
ditto `isnull()`

san7890 commented on code/modules/mob/living/carbon/human/human_update_icons.dm:
```diff
@@ -295,9 +296,17 @@ There are several things that need to be remembered:
+		var/feature_y_offset = 0
+		for (var/body_zone in list(BODY_ZONE_L_LEG, BODY_ZONE_R_LEG))
+			var/obj/item/bodypart/leg/my_leg = get_bodypart(body_zone)
+			if(!my_leg)
```
ditto isnull

san7890 commented on code/modules/mob/living/carbon/human/human_update_icons.dm:
```diff
@@ -421,7 +426,8 @@ There are several things that need to be remembered:
 
-	if(!get_bodypart(BODY_ZONE_HEAD)) //Decapitated
+	var/obj/item/bodypart/head/my_head = get_bodypart(BODY_ZONE_HEAD)
+	if(!my_head) //Decapitated
```
ditto isnull

san7890 commented on code/modules/surgery/organs/vocal_cords.dm:
```diff
@@ -120,3 +85,38 @@
+
+/obj/item/organ/internal/vocal_cords/adamantine/handle_speech(message)
+	var/msg = span_resonate(span_name("[owner.real_name]</span> <span class='message'>resonates, \"[message]\""))
+	for(var/player in GLOB.player_list)
```
this probably isn't a "choke out the server" level proc but should we have a `CHECK_TICK` in this for loop?

LemonInTheDark replied:
it's alright, this is basically what say() already does for ghosts and all. It'd be about 90 iterations at worst, but the actual work is pretty light. `get_organ_slot` is just a fancy list read (w proc overhead but it's ok)

san7890 commented on code/modules/surgery/organs/vocal_cords.dm:
```diff
@@ -120,3 +85,38 @@
+	owner.say(".x[message]")
+
+/obj/item/organ/internal/vocal_cords/adamantine/handle_speech(message)
+	var/msg = span_resonate(span_name("[owner.real_name]</span> <span class='message'>resonates, \"[message]\""))
```
this is yonked as shit and `<span class='message'>` doesn't even seem to do anything according to my precursory look through the span classes tgui-side? let's just do this.

```suggestion
	var/msg = (span_resonate("[span_name("[owner.real_name]")] resonates, \"[message]\""))
```

san7890 replied:
honestly there not being a span class for "message" (whatever that means) is probably why the greps didn't catch this two years ago

san7890 commented on code/modules/mining/ores_coins.dm:
```diff
@@ -260,52 +261,60 @@ GLOBAL_LIST_INIT(sand_recipes, list(\
+	playsound(src,'sound/effects/hit_on_shattered_glass.ogg',50,TRUE)
+	icon_state = "Gibtonite active"
+	var/notify_admins = FALSE
+	if(z != 5)//Only annoy the admins ingame if we're triggered off the mining zlevel
```
you don't need to change this now but WOW this is so prone to breaking (in fact it might already be broken on northstar), we really need to read the ZTRAITS instead.

LemonInTheDark commented on code/__DEFINES/atmospherics/atmos_mob_interaction.dm:
```diff
@@ -59,10 +59,14 @@
 /// This also affects how fast the body normalises it's temperature when hot.
 /// 340k is about 66c, and rather high for a human.
 #define BODYTEMP_HEAT_DAMAGE_LIMIT (BODYTEMP_NORMAL + 30)
+/// A temperature limit which is above the maximum lavaland temperature
+#define BODYTEMP_HEAT_LAVALAND_SAFE (BODYTEMP_HEAT_DAMAGE_LIMIT + 20)
 /// The body temperature limit the human body can take before it starts taking damage from cold.
```
if you're going to make this a define, could you ensure it's tied to atmosphere generation figures somehow? or at least throw an error if it goes below that threshold

LemonInTheDark commented on code/__DEFINES/atmospherics/atmos_mob_interaction.dm:
```diff
@@ -59,10 +59,14 @@
 /// 270k is about -3c, that is below freezing and would hurt over time.
 #define BODYTEMP_COLD_DAMAGE_LIMIT (BODYTEMP_NORMAL - 40)
+/// A temperature limit which is above the minimum icebox temperature
+#define BODYTEMP_COLD_ICEBOX_SAFE (BODYTEMP_COLD_DAMAGE_LIMIT - 40)
```
ditto

Jacquerel committed:
The trivial fixes

Jacquerel committed:
More defines for planetary temperature, much more sensible actually
Fixes funny hand offsets

Jacquerel committed:
Is this a harddel? <= clueless

Jacquerel committed:
oh lol i was using the wrong qdel_list
cosmetic things with eyelessness and brainlessness

Jacquerel committed:
Merge remote-tracking branch 'tgstation/master' into subtypes-begone

Overuseddart commented:
Wait where is the cool syndie sprite?

Overuseddart commented:
What’s the explosion radius and damage for the gibonite explosion?

Jacquerel commented:
> What’s the explosion radius and damage for the gibonite explosion?

Depends on the gibtonite, it just works like gibtonite. If that makes it too wild after experimenting with it, then we'll change it.

Overuseddart commented:
> Depends on the gibtonite, it just works like gibtonite. If that makes it too wild after experimenting with it, then we'll change it.

I’m pretty sure gibonite can put key targets like the Captain the HoS, and secoffs into crit with one hit. They can put round start miners with 1-2 Goliath plates into crit reliably and even delimb them.

Jacquerel committed:
Merge branch 'master' into subtypes-begone

MrMelbert reviewed (changes requested):
I kind of glossed over the bodypart section, lots of stuff going around, wonder how the icon cache takes it

Jacquerel commented on code/game/atoms.dm:
```diff
@@ -375,6 +375,7 @@
 /atom/proc/handle_ricochet(obj/projectile/ricocheting_projectile)
 	var/turf/p_turf = get_turf(ricocheting_projectile)
 	var/face_direction = get_dir(src, p_turf)
+	face_direction = face_direction ? face_direction : get_dir(src, ricocheting_projectile)
```
Yeah it returns 0, real pain in the ass.

Jacquerel committed:
Some syntax stuff
Remove unused proc
Use some helpers
Little bit of memory management

Jacquerel commented:
> I kind of glossed over the bodypart section, lots of stuff going around, wonder how the icon cache takes it

It's nothing that changes on a _per-instance_ level rather than a per-subtype level so afaik it should be fine?

san7890 commented on code/modules/surgery/organs/vocal_cords.dm:
```diff
@@ -120,3 +85,38 @@
+	owner.say(".x[message]")
+
+/obj/item/organ/internal/vocal_cords/adamantine/handle_speech(message)
+	var/msg = span_resonate(span_name("[owner.real_name] resonates, \"[message]\""))
```
```suggestion
	var/msg = span_resonate([span_name("[owner.real_name]")] resonates, \"[message]\""))
```

you changed the meaning of the spans such that `span_name` would apply to the entire string, not just the actual name portion. it's not your fault because the last one was so god damn confusing but i'm 99% sure this is what was meant

Jacquerel committed:
Fix formatting

Jacquerel committed:
Merge branch 'master' into subtypes-begone

Jacquerel committed:
Merge branch 'master' into subtypes-begone

Jacquerel committed:
Merge branch 'master' into subtypes-begone

Jacquerel committed:
Move action deletion to Destroy

Jacquerel committed:
Oopsie

SinfulBliss commented:
does this include metal h2 golems?

nukularpower commented:
Gotta say I am not a fan of this at all so far.   The immobility after not eating is something that should not exist in my mind - becoming very slow sure, but the final freezup into death (unless you've raided a techfab) is just not fun.   Some of the forms are basically useless or have nothing special or thematic about them, such as Uranium, while others seem broken, like plasma form being completely invulnerable to lasers/heat - that made killing a blob a joke with an emitter to heal me up In my last golem game, not to mention being immune to almost all station ranged weaponry.   I hope runic metal and metal hydrogen forms will be added as well.

 In the round I mentioned, something weird happened where I could not fill up after eating plasma - even though I was continually chomping it, I still died of starvation.   I could not try eating anything else due to being locked to plasma form - it feels like there should be a way to voluntarily "discharge" your current form, even if on a long cooldown.

All of that aside, what I really came to report is that the new golem form seems to have trouble punching many things, most importantly including blob tiles - it cannot hit them with a bare hand.

optimumtact commented:
Glass is the omni food not iron in my design doc but you have to eat a lot of it.

This mechanic probably needs to prevent you eating other types of materials that will have no effect otherwise it's probably a bit hard to tell what's going on.

As for the blob thing, that's on the blob for not swapping forms there's always counter plays to the forms you take.


Also it's kind of weird you both argue that this golem should be buffed (by removing the primary downside) and then also in the same breath argue it's overpowered.

nukularpower commented:
Hmm, I certainly didnt mean to imply plasma golem should be buffed - it is quite strong and I would indeed call it overpowered.  Other forms could use buffs or more uniqueness in my eyes, like uranium, plasteel, etc.  Missed that it is glass to feed on, not iron.

I'm not sure what the blob could have done by changing forms - I was a plasma golem so healed from heat damage, and there was an emitter I stood in front of that kept me at full HP while I used a crowbar to beat the blob to death (since again, golems cant punch blob tiles for some reason.)  The blob did change forms, but even with a blobbernaut beating me and the blob spamming melee attacks, they just couldnt do enough damage to hurt me through the self healing from emitter and sec's laser spam.  Again, that all feels redundant with the effect of gold as well.

Jacquerel commented:
There is certainly some redundancy with gold but I think you have misunderstood the effect of plasma a little, that emitter wasn't healing you (and I have booted up a test instance just to double check I didn't fuck up somewhere) it just also wasn't really hurting you. Not sure exactly why in that case you felt so durable against blob attacks, but it does seem like probably the blob could have done something about it.

I'll look into the bug with eating a whole stack at once, that shouldn't happen.
And potentially the bananium one too, as while I agree that it's funnier that way we should probably stick to the design doc for a while before making changes. Tactical resting is an honoured clown technique.

CCC23 commented:
telecrystal golem when?

Elli-Skala commented:
> telecrystal golem when?

This pr is two years old
